File: ios/Arculus/CSDK/WalletGetGGUID.swift
import CSDK

final class WalletGetGGUID: CSDKAPICall<String> {
    override func request() async throws -> [Data] {
        var len: size_t = 0
        
        guard let bytes = WalletGetGGUIDRequest(wallet, &len) else {
            throw CSDKAPICallError(.requestCreationFailed("wallet get GGUID"))
        }
        
        return [Data(bytes: bytes, count: len)]
    }
    
    override func response(bytes: [UInt8]) async throws -> ResponseType {
        var len: size_t = 0
        
        guard let result = WalletGetGGUIDResponse(wallet, bytes, bytes.count, &len) else {
            throw CSDKAPICallError(.responseParsingFailed("wallet get GGUID"))
        }
        
        let data = Data(bytes: result, count: len)
        
        return dataToHexString(data)
    }
}

File: android/src/main/java/com/aptoslabs/arculus/csdk/WalletGetGGUID.kt
package com.aptoslabs.arculus.csdk

import com.sun.jna.Pointer

class WalletGetGGUID(wallet: Pointer) : CSDKAPICall<String>(wallet) {
  override suspend fun request(): Array<ByteArray?> {
    val len = SizeTByReference()

    val pointer = CSDK.WalletGetGGUIDRequest(wallet, len)

    if (pointer == Pointer.NULL) throw CSDKAPICallRequestCreationException()

    return arrayOf(pointer.getByteArray(0, len.toIntChecked()))
  }

  override suspend fun response(bytes: ByteArray): String {
    val len = SizeTByReference()

    val pointer = CSDK.WalletGetGGUIDResponse(wallet, bytes, bytes.size, len)

    if (pointer == Pointer.NULL) throw CSDKAPICallResponseParsingException()

    val gguidBytes = pointer.getByteArray(0, len.toIntChecked())

    return byteArrayToHexString(gguidBytes)
  }
}
